### ISSUE TITLE
Broken selinux.mode for RHEL6

### ISSUE THREAD

**Issue by atyagi1024** [saltstack/salt#26703]
While trying to force selinux.mode RHEL6 keeps throwing the below error. Is there anything that I may have missed?
```python-traceback
permissive:
    selinux.mode

[ERROR   ] An un-handled exception was caught by salt's global exception handler:
KeyError: 'selinux.mode'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 681, in sls
    ret = st_.state.call_high(high_)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'selinux.mode'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 227, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 69, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 236, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 681, in sls
    ret = st_.state.call_high(high_)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2067, in call_high
    ret = dict(list(disabled.items()) + list(self.call_chunks(chunks).items()))
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1623, in call_chunks
    running = self.call_chunk(low, running, chunks)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 1769, in call_chunk
    self._mod_init(low)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 612, in _mod_init
    self.states['{0}.{1}'.format(low['state'], low['fun'])]  # pylint: disable=W0106
  File "/usr/lib/python2.6/site-packages/salt/utils/lazy.py", line 90, in __getitem__
    raise KeyError(key)
KeyError: 'selinux.mode'
```
```yaml
[root@ldaplabn1 ~]# salt-call --versions-report
           Salt: 2015.5.3
         Python: 2.6.6 (r266:84292, Nov 21 2013, 10:50:32)
         Jinja2: unknown
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.5
           Mako: Not Installed
        Tornado: Not Installed
[root@ldaplabn1 ~]# 
[root@ldaplabn1 ~]# cat /etc/redhat-release 
Red Hat Enterprise Linux Server release 6.6 (Santiago)
```

**Comment by jfindlay** [comment]
@atyagi1024, thanks for the report.  What does `salt-call sys.doc selinux.getenforce` return?  If it returns nothing, then it may be that salt thinks selinux is not installed.

**Comment by atyagi1024** [comment]
Yes, nothing gets return for salt-call sys.doc selinux.getenforce. Also please note that it is a problem on RHEL5 as well.

**Comment by lorengordon** [comment]
To use manage selinux with python, we found that we had to ensure the selinux policy package was installed (`selinux-policy-targeted`), as well as the python module for managing selinux (`policycoreutils-python`). You can install both via yum.

**Comment by atyagi1024** [comment]
thanks, policycoreutils-python was missing. Installing that fixed the issue on RHEL6. But RHEL5 has the same problem as well. I have the following packages installed. Anything package that I might be missing on RHEL5?

rpm -qa | egrep "selinux|policycoreutils"
===============================
libselinux-1.33.4-5.7.el5
libselinux-utils-1.33.4-5.7.el5
policycoreutils-1.33.12-14.13.el5
selinux-policy-devel-2.4.6-351.el5
libselinux-python-1.33.4-5.7.el5
selinux-policy-minimum-2.4.6-351.el5
selinux-policy-strict-2.4.6-351.el5
selinux-policy-2.4.6-351.el5
libselinux-1.33.4-5.7.el5
selinux-policy-targeted-2.4.6-351.el5
policycoreutils-newrole-1.33.12-14.13.el5
selinux-policy-mls-2.4.6-351.el5

**Comment by lorengordon** [comment]
Well, I don't see `policycoreutils-python` in that list, so maybe the RHEL5 instance is still missing that one?

**Comment by atyagi1024** [comment]
Sorry but I couldn't find policycoreutils-python for RHEL5.

**Comment by lorengordon** [comment]
It may be that the __virtual__() function in salt's selinux module is preventing it from loading. I don't have any RHEL5 instances to test on, but if you're comfortable reviewing the code, you should be able to figure out what it's checking for and compare that against your RHEL5 system to see what might be missing.

https://github.com/saltstack/salt/blob/develop/salt/modules/selinux.py#L27

**Comment by jfindlay** [comment]
@lorengordon, that should be https://github.com/saltstack/salt/blob/2015.5.3/salt/modules/selinux.py#L27.

**Comment by jfindlay** [comment]
It looks like to me that the selinux checks for salt don't account for rhel5-specific features and need to be fixed for that case.

**Comment by lorengordon** [comment]
@jfindlay, side question, do you know, is there a way to associate a dependency to a module/state? I know you probably don't want to add `policycoreutils-python` as a hard dependency for salt, but it would be nice if we could associate the os:dependency-map to the module/state so that salt could check for it and output a nice error message if any necessary dependencies were missing.

**Comment by drawsmcgraw** [comment]
@lorengordon It does seem to be that the virtual function is refusing to load it (at least in my case).

I'm having the same issue here even after installing `selinux-policy-targeted` and `policycoreutils-python`. This is on CentOS 6.4.

After looking at [the `selinux_fs_path()` function that the virtual function in the selinux module calls](https://github.com/saltstack/salt/blob/develop/salt/modules/selinux.py#L59-L62), I seem to lack the file that the function `selinux_fs_path()` is looking for.

```
[root@cm ~]# ls -la /sys/fs/
total 0
drwxr-xr-x  3 root root 0 Aug 28 12:20 .
drwxr-xr-x 13 root root 0 Jul 27 08:59 ..
drwxr-xr-x  5 root root 0 Aug 28 12:26 ext4

[root@cm ~]# ls -la /selinux/
total 8
drwxr-xr-x.  2 root root 4096 Apr 18  2013 .
dr-xr-xr-x. 24 root root 4096 Jul 30 13:29 ..
```

To be fair, SELinux was never enabled on these machines. I just want to keep it that way.

Obligatory versions-report:

```
[root@edh-dev-cm ~]# salt-call --versions-report
           Salt: 2015.5.3
         Python: 2.6.6 (r266:84292, Jul 23 2015, 15:22:56)
         Jinja2: 2.2.1
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.5
           Mako: Not Installed
        Tornado: Not Installed
```

**Comment by jfindlay** [comment]
@lorengordon, that is a good question.  I don't know if we have anything like that.  I think it would be great if each plugin (exec module, state module, runner, beacon, etc.) that needed external dependencies could plug that info into a central aggregator that could return the full set of extended dependencies depending on what modules are loaded, or potentially all modules, so that we could keep `salt --versions-report` exclusive to core dependencies.  If you want, you could open a new issue for this.

**Comment by lorengordon** [comment]
@drawsmcgraw, interesting. I ran into that same problem with Amazon Linux, but it wasn't critical and I brushed it off to investigate later. So, yeah, there are situations where those paths don't exist, even when all other dependencies are met. Not sure what a good alternate test would be for the __virtual()__ function...

**Comment by lorengordon** [comment]
@jfindlay, done, https://github.com/saltstack/salt/issues/26755.

**Comment by drawsmcgraw** [comment]
@lorengordon Fair enough. I already have `/etc/selinux/config` being managed with Salt, I was just being aggressive with also setting the `mode` to 0. I've just commented that out for now; not a huge impact for me. Wish I had a little more time to investigate that :/

**Comment by emddudley** [comment]
I encountered the `KeyError: 'selinux.mode'` problem using Salt 2015.5.5 on CentOS 7. Installing the `policycoreutils-python` package was sufficient to resolve it. The `selinux-policy-targeted` package was already installed on my machine since it's part of the `Core` group.

However I did run into a related issue. I thought it would be clever to add a state that would install the required packages:

    permissive:
      selinux.mode:
        - require:
          - pkg: policy-packages
    
    policy-packages:
      pkg.installed:
        - pkgs:
          - policycoreutils-python
          - selinux-policy-targeted

This did *not* work. Something to do with the order in which states are parsed and modules are loaded? The reason why is probably obvious to people who are familiar with Salt's internals, but I am not. Anyways, swapping those states around so that the `policy-packages` state is listed first in my SLS *does* work. I don't like that the declaration order matters here, but I want my Salt states to execute successfully without having to pre-install dependencies.

**Comment by drawsmcgraw** [comment]
@emddudley Interesting... I could swear things were still broken for me when even after installing those deps.

> I don't like that the declaration order matters here, but I want my Salt states to execute successfully without having to pre-install dependencies.

If you want to disable the top-down ordering in Salt (I wouldn't recommend it but I'm only one person), you can set `state_auto_order` to `False` in the Master config. That will cause Salt to order the states in a lexicographic manner instead of top-down as defined in the files. You can see the [Ordering States](https://docs.saltstack.com/en/latest/ref/states/ordering.html) docs for more details.

**Comment by emddudley** [comment]
@drawsmcgraw I am defining my states declaratively so that their order in my SLS files doesn't matter. I just want Salt to honor my requisites. And that's where I have the problem: the `KeyError` is triggered when my states are out-of-order as laid out in my SLS, even when the requisites should be forcing them to execute in the proper order.

It might be difficult or impossible to change this behavior, I don't really know. I am just noting that it is inconsistent for state errors to occur based on one state ordering (auto or lexicographic) while the states would actually be executing in a different order (requisites).

**Comment by drawsmcgraw** [comment]
@emddudley Ah, I see. My apologies. I wasn't paying attention and missed the requisites you're using. I agree, that's not the correct behavior.

**Comment by rallytime** [comment]
The originally reported stacktrace was caused by a regression in 2015.5.3 with the state compiler. There were several other bug displaying this problem such as #27619, #26882, #26454, #27016, and #26280 and should be fixed 2015.5.6 and 2015.8.0. Can anyone here test and confirm this either at one of those version tags or at the HEAD of the 2015.8 or 2015.5 branch?

The backstory here is that this bug was fixed on the 2015.8 branch with #27681, which I back-ported to 2015.5 with #27681.

**Comment by rallytime** [comment]
Update: It appears that I mis-remembered when my back-porting PR went into the 2015.5 branch and it missed the 2015.5.6 release tag. However, the initial stacktrace reported by @atyagi1024 (and those seen by @lorengordon, @drawsmcgraw, and @emddudley) about the KeyError should be fixed at the HEAD of the 2015.5 branch. The statement about this fix being in 2015.8.0 still stands.

**Comment by rallytime** [comment]
Since this regression has been fixed on the 2015.5 branch and the fix will be present on the 2015.5.7 release of Salt (and is also fixed in 2015.8.0 and newer), I am closing this issue. If people see problems here again, let us know and we will be happy to look at this issue again. Thanks!

**Comment by vans163** [comment]
This recipe for me on Fedora 24 gives error:
```
#kernel.sls
kernel_selinux:
  pkg.installed:
    - pkgs:
      - policycoreutils
      - policycoreutils-python
  selinux.mode:
    - name: disabled


----------
          ID: kernel_selinux
    Function: pkg.installed
      Result: True
     Comment: All specified packages are already installed
     Started: 17:00:25.891519
    Duration: 872.757 ms
     Changes:   
----------
          ID: kernel_selinux
    Function: selinux.mode
        Name: disabled
      Result: False
     Comment: State 'selinux.mode' was not found in SLS 'kernel'
              Reason: 'selinux' __virtual__ returned False
     Started: 
    Duration: 
     Changes: 

```
